### DISCUSSION TITLE
AllocationLimitMegabytes seems not to be followed with GIF's

### DISCUSSION THREAD

**Discussion by ErazerBrecht** [SixLabors/ImageSharp#2799] [Q&A] [answered]
Before creating an actual issue I wanted to check if I'm missing something crucial.

A while ago I learnt from James that you can set a limit of the max allocation for an image.
https://github.com/SixLabors/ImageSharp/issues/2758#issuecomment-2194834465

I was playing with this and one of my co-workers created a GIF that still allocates 6Gb of RAM even while I configured the limit on 192Mb.

POC: https://github.com/ErazerBrecht/ImageSharp.Payloads/blob/master/ImageSharp.Payloads.GifMemory/Program.cs

Proof:
![image](https://github.com/user-attachments/assets/e521fa13-efbc-46aa-81f2-a488042f22fe)

I guess this works per frame?
Someone has an idea on how to make sure to limit memory allocation for GIF's?

Or is this a bug and should I create an issue?

Sincerely,
Brecht

**Comment by antonfirsov** [comment, marked as answer]
`AllocationLimitMegabytes` only limit a single allocation while gif-s are multi frame / multi buffer objects.

My recommendation would be to either set [`DecoderOptions.MaxFrames`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.DecoderOptions.html#SixLabors_ImageSharp_Formats_DecoderOptions_MaxFrames), or use `Image.Identify(Async)` before decoding to set a limit for the overall image allocation before decoding:

```C#
int totalImageLimitInBytes = 192 * 1024 * 1024;
ImageInfo info = await Image.IdentifyAsync(...);
int frameCount = info.FrameMetadataCollection.Count;

if (info.Width * info.Height * frameCount * sizeof(Rgba32) > totalImageLimitInBytes)
{
    throw new Exception("Image too large.");
}
```

Note: we shall probably consider a built-in feature to achieve the above.

**Reply by ErazerBrecht** [reply]
Okay as expected, thanks @antonfirsov for confirming.
A check like that we already have in our production code.

We were just checking if this global limit could help us even harder in some other edge cases.
Glad to hear you are considering a built-in feature!

Enjoy the rest of the day and thanks again!

Brecht

**Reply by JimBobSquarePants** [reply]
I’m still going to double check the image to make sure we haven’t missed something.

**Reply by JimBobSquarePants** [reply]
@ErazerBrecht  OK. Nothing to see here. 40000 * 1000 * 41 frames @ 4 bytes per pixel = 6560000000 bytes so it's just a massive image.

@antonfirsov if we were to ever do something like that, how would you track current allocations?
